File: 0853-car-fleet/0853-car-fleet.java
class Solution {
    class Car implements Comparable<Car>{
        int position;
        int speed;
        double time;
        public Car(int p, int s, double t){
            this.position = p;
            this.speed = s;
            this.time = t;
        }
        @Override
        public int compareTo(Car c){
            return this.position - c.position;
        }
    }
    public int carFleet(int target, int[] position, int[] speed) {
        int n = position.length;
        if(n == 1 || n == 0) return n;
        List<Car> allCars = new ArrayList<>();
        for(int i = 0; i < n; i ++) allCars.add(new Car(position[i], speed[i], (double)(target - position[i])/speed[i]));
        Collections.sort(allCars);
        int totalFleets = 1;
        double curFleetTime = allCars.get(n-1).time;
        for(int i = allCars.size()-2; i >= 0; i--){
            if(allCars.get(i).time > curFleetTime){
                curFleetTime = allCars.get(i).time;
                totalFleets++;
            }
        }
        return totalFleets;
    }
}

File: 0853-car-fleet/0853-car-fleet.py
class Solution:
    class Car:
        def __init__(self, position, speed, time):
            self.position = position
            self.speed = speed
            self.time = time

        def __lt__(self, other):
            return self.position < other.position
        
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        n = len(position)
        if n == 1 or n == 0:
            return n

        all_cars = [self.Car(position[i], speed[i], (target - position[i]) / speed[i]) for i in range(n)]
        all_cars.sort()

        total_fleets = 1
        cur_fleet_time = all_cars[-1].time

        for i in range(n - 2, -1, -1):
            if all_cars[i].time > cur_fleet_time:
                cur_fleet_time = all_cars[i].time
                total_fleets += 1

        return total_fleets
        